{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython\n",
        "!apt-get update\n",
        "!apt-get install -y ncbi-blast+\n",
        "\n",
        "#  Installs BLAST+ tools and BioPython"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EUAsw4Uezcey",
        "outputId": "1579a82b-9b04-44da-e91c-a4a4312b3f0c"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[?25l   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/3.3 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[91m╸\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m122.4 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m64.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.85\n",
            "Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Hit:3 https://cli.github.com/packages stable InRelease\n",
            "Get:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Get:6 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]\n",
            "Get:7 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Get:8 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n",
            "Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Get:10 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,942 kB]\n",
            "Hit:11 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:12 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [5,543 kB]\n",
            "Get:14 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages [80.2 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,576 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,617 kB]\n",
            "Get:17 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,783 kB]\n",
            "Get:18 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,209 kB]\n",
            "Get:19 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [5,340 kB]\n",
            "Get:20 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,272 kB]\n",
            "Get:21 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [3,297 kB]\n",
            "Fetched 35.1 MB in 4s (9,111 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  ncbi-data\n",
            "The following NEW packages will be installed:\n",
            "  ncbi-blast+ ncbi-data\n",
            "0 upgraded, 2 newly installed, 0 to remove and 41 not upgraded.\n",
            "Need to get 15.8 MB of archives.\n",
            "After this operation, 71.8 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ncbi-data all 6.1.20170106+dfsg1-9 [3,519 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ncbi-blast+ amd64 2.12.0+ds-3build1 [12.3 MB]\n",
            "Fetched 15.8 MB in 1s (11.4 MB/s)\n",
            "Selecting previously unselected package ncbi-data.\n",
            "(Reading database ... 126371 files and directories currently installed.)\n",
            "Preparing to unpack .../ncbi-data_6.1.20170106+dfsg1-9_all.deb ...\n",
            "Unpacking ncbi-data (6.1.20170106+dfsg1-9) ...\n",
            "Selecting previously unselected package ncbi-blast+.\n",
            "Preparing to unpack .../ncbi-blast+_2.12.0+ds-3build1_amd64.deb ...\n",
            "Unpacking ncbi-blast+ (2.12.0+ds-3build1) ...\n",
            "Setting up ncbi-data (6.1.20170106+dfsg1-9) ...\n",
            "Setting up ncbi-blast+ (2.12.0+ds-3build1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!blastdbcmd -version\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_jiDtGDTzhZa",
        "outputId": "7ae8402a-41d1-40ca-819e-a538be622afe"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "blastdbcmd: 2.12.0+\n",
            " Package: blast 2.12.0, build Mar  8 2022 16:19:08\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "import os\n",
        "import subprocess\n",
        "import sqlite3\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "import re\n",
        "from google.colab import files"
      ],
      "metadata": {
        "id": "YOC2jEcyzjJs"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Upload Your Downloaded Files"
      ],
      "metadata": {
        "id": "ueMUqmOn9i87"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "print(\"Upload your BLAST database files:\")\n",
        "print(\"You need to upload:\")\n",
        "print(\"- All 18S_fungal_sequences.* files (ndb, nhr, nin, nsq, etc.)\")\n",
        "print(\"- taxdb.btd\")\n",
        "print(\"- taxdb.bti\")\n",
        "print(\"- taxonomy4blast.sqlite3\")\n",
        "\n",
        "# Upload files\n",
        "uploaded = files.upload()\n",
        "\n",
        "print(f\"Uploaded files: {list(uploaded.keys())}\")\n",
        "\n",
        "# List all files in current directory\n",
        "print(\"\\nAll files in directory:\")\n",
        "for file in os.listdir(\".\"):\n",
        "    size = os.path.getsize(file) / 1024  # KB\n",
        "    print(f\"  {file} ({size:.1f} KB)\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 995
        },
        "id": "4LUWqhpxzyT7",
        "outputId": "c1a81f5d-2d07-4ff8-be3a-852bdc11312f"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Upload your BLAST database files:\n",
            "You need to upload:\n",
            "- All 18S_fungal_sequences.* files (ndb, nhr, nin, nsq, etc.)\n",
            "- taxdb.btd\n",
            "- taxdb.bti\n",
            "- taxonomy4blast.sqlite3\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-9881a046-7cc6-4601-9cb9-e5462f67cd2b\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-9881a046-7cc6-4601-9cb9-e5462f67cd2b\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving 18S_fungal_sequences.ndb to 18S_fungal_sequences (1).ndb\n",
            "Saving 18S_fungal_sequences.nhr to 18S_fungal_sequences.nhr\n",
            "Saving 18S_fungal_sequences.nin to 18S_fungal_sequences.nin\n",
            "Saving 18S_fungal_sequences.nnd to 18S_fungal_sequences.nnd\n",
            "Saving 18S_fungal_sequences.nni to 18S_fungal_sequences.nni\n",
            "Saving 18S_fungal_sequences.nog to 18S_fungal_sequences.nog\n",
            "Saving 18S_fungal_sequences.nos to 18S_fungal_sequences.nos\n",
            "Saving 18S_fungal_sequences.not to 18S_fungal_sequences.not\n",
            "Saving 18S_fungal_sequences.nsq to 18S_fungal_sequences.nsq\n",
            "Saving 18S_fungal_sequences.ntf to 18S_fungal_sequences.ntf\n",
            "Saving 18S_fungal_sequences.nto to 18S_fungal_sequences.nto\n",
            "Saving taxdb.btd to taxdb.btd\n",
            "Saving taxdb.bti to taxdb.bti\n",
            "Saving taxonomy4blast.sqlite3 to taxonomy4blast.sqlite3\n",
            "Uploaded files: ['18S_fungal_sequences (1).ndb', '18S_fungal_sequences.nhr', '18S_fungal_sequences.nin', '18S_fungal_sequences.nnd', '18S_fungal_sequences.nni', '18S_fungal_sequences.nog', '18S_fungal_sequences.nos', '18S_fungal_sequences.not', '18S_fungal_sequences.nsq', '18S_fungal_sequences.ntf', '18S_fungal_sequences.nto', 'taxdb.btd', 'taxdb.bti', 'taxonomy4blast.sqlite3']\n",
            "\n",
            "All files in directory:\n",
            "  .config (4.0 KB)\n",
            "  18S_fungal_sequences.nos (73.6 KB)\n",
            "  taxdb.bti (17621.7 KB)\n",
            "  taxdb.btd (167178.8 KB)\n",
            "  18S_fungal_sequences.nog (14.5 KB)\n",
            "  18S_fungal_sequences.nto (29.8 KB)\n",
            "  18S_fungal_sequences.nhr (627.4 KB)\n",
            "  18S_fungal_sequences.nni (0.2 KB)\n",
            "  18S_fungal_sequences.nin (43.4 KB)\n",
            "  18S_fungal_sequences.ntf (140.0 KB)\n",
            "  18S_fungal_sequences.not (43.8 KB)\n",
            "  18S_fungal_sequences (1).ndb (224.0 KB)\n",
            "  18S_fungal_sequences.nsq (1207.2 KB)\n",
            "  18S_fungal_sequences.ndb (224.0 KB)\n",
            "  taxonomy4blast.sqlite3 (84248.0 KB)\n",
            "  18S_fungal_sequences.nnd (29.8 KB)\n",
            "  sample_data (4.0 KB)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def convert_blast_to_fasta(db_name=\"18S_fungal_sequences\", output_file=\"18S_sequences.fasta\"):\n",
        "    \"\"\"Convert BLAST database to FASTA format\"\"\"\n",
        "\n",
        "    print(f\"Converting {db_name} to FASTA format...\")\n",
        "\n",
        "    # Check if required database files exist\n",
        "    required_extensions = ['.ndb', '.nhr', '.nin', '.nsq']\n",
        "    required_files = [f\"{db_name}{ext}\" for ext in required_extensions]\n",
        "\n",
        "    print(\"Checking for required database files:\")\n",
        "    missing_files = []\n",
        "    for file in required_files:\n",
        "        if os.path.exists(file):\n",
        "            print(f\"  ✓ {file}\")\n",
        "        else:\n",
        "            print(f\"  ❌ {file} (missing)\")\n",
        "            missing_files.append(file)\n",
        "\n",
        "    if missing_files:\n",
        "        print(f\"\\n❌ Missing files: {missing_files}\")\n",
        "        print(\"Please upload all the 18S_fungal_sequences.* files\")\n",
        "        return False\n",
        "\n",
        "    try:\n",
        "        # Run blastdbcmd to convert to FASTA\n",
        "        print(f\"\\nRunning blastdbcmd...\")\n",
        "\n",
        "        result = subprocess.run([\n",
        "            \"blastdbcmd\",\n",
        "            \"-db\", db_name,\n",
        "            \"-entry\", \"all\",\n",
        "            \"-outfmt\", \"%f\",  # FASTA format\n",
        "            \"-out\", output_file\n",
        "        ], capture_output=True, text=True, check=True)\n",
        "\n",
        "        print(f\"✓ Conversion successful!\")\n",
        "        print(f\"STDOUT: {result.stdout}\")\n",
        "\n",
        "        # Check output\n",
        "        if os.path.exists(output_file):\n",
        "            file_size = os.path.getsize(output_file) / (1024*1024)  # MB\n",
        "            print(f\"✓ Created {output_file} ({file_size:.2f} MB)\")\n",
        "\n",
        "            # Quick count of sequences\n",
        "            with open(output_file, 'r') as f:\n",
        "                seq_count = sum(1 for line in f if line.startswith('>'))\n",
        "            print(f\"✓ Contains {seq_count} sequences\")\n",
        "\n",
        "            return True\n",
        "        else:\n",
        "            print(\"❌ Output file not created\")\n",
        "            return False\n",
        "\n",
        "    except subprocess.CalledProcessError as e:\n",
        "        print(f\"❌ blastdbcmd failed!\")\n",
        "        print(f\"Error: {e}\")\n",
        "        print(f\"STDERR: {e.stderr}\")\n",
        "        return False\n",
        "    except Exception as e:\n",
        "        print(f\"❌ Unexpected error: {e}\")\n",
        "        return False\n",
        "\n",
        "# Run conversion\n",
        "conversion_success = convert_blast_to_fasta()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fPN10vSx6FHD",
        "outputId": "4387d825-0c23-4951-a453-8259518880a9"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Converting 18S_fungal_sequences to FASTA format...\n",
            "Checking for required database files:\n",
            "  ✓ 18S_fungal_sequences.ndb\n",
            "  ✓ 18S_fungal_sequences.nhr\n",
            "  ✓ 18S_fungal_sequences.nin\n",
            "  ✓ 18S_fungal_sequences.nsq\n",
            "\n",
            "Running blastdbcmd...\n",
            "✓ Conversion successful!\n",
            "STDOUT: \n",
            "✓ Created 18S_sequences.fasta (5.12 MB)\n",
            "✓ Contains 3692 sequences\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Load and Inspect FASTA Sequence"
      ],
      "metadata": {
        "id": "F_nGvNtx_jOS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def load_and_inspect_fasta(fasta_file=\"18S_sequences.fasta\", sample_size=20):\n",
        "    \"\"\"Load FASTA file and show detailed inspection\"\"\"\n",
        "\n",
        "    if not os.path.exists(fasta_file):\n",
        "        print(f\"❌ {fasta_file} not found!\")\n",
        "        return None, None, None\n",
        "\n",
        "    print(f\"Loading sequences from {fasta_file}...\")\n",
        "\n",
        "    sequences = []\n",
        "    seq_ids = []\n",
        "    descriptions = []\n",
        "\n",
        "    # Load all sequences\n",
        "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
        "        sequences.append(str(record.seq))\n",
        "        seq_ids.append(record.id)\n",
        "        descriptions.append(record.description)\n",
        "\n",
        "    print(f\"✓ Loaded {len(sequences)} sequences\")\n",
        "\n",
        "    # Detailed inspection\n",
        "    print(f\"\\n🔍 DETAILED INSPECTION (First {sample_size} sequences):\")\n",
        "    print(\"=\" * 100)\n",
        "\n",
        "    for i in range(min(sample_size, len(sequences))):\n",
        "        print(f\"\\n[{i+1}] SEQUENCE ID: {seq_ids[i]}\")\n",
        "        print(f\"DESCRIPTION: {descriptions[i]}\")\n",
        "        print(f\"LENGTH: {len(sequences[i])} bp\")\n",
        "        print(f\"SEQUENCE: {sequences[i][:80]}{'...' if len(sequences[i]) > 80 else ''}\")\n",
        "\n",
        "        # Try to extract taxonomy from description\n",
        "        genus_species = re.search(r'\\b([A-Z][a-z]+)\\s+([a-z]+)\\b', descriptions[i])\n",
        "        if genus_species:\n",
        "            print(f\"🏷️  EXTRACTED TAXA: {genus_species.group(1)} {genus_species.group(2)}\")\n",
        "        else:\n",
        "            # Look for any capitalized words\n",
        "            taxa_words = re.findall(r'\\b[A-Z][a-z]{3,}\\b', descriptions[i])\n",
        "            if taxa_words:\n",
        "                print(f\"🏷️  POTENTIAL TAXA: {taxa_words[:3]}\")\n",
        "            else:\n",
        "                print(\"🏷️  NO CLEAR TAXA FOUND\")\n",
        "\n",
        "        print(\"-\" * 80)\n",
        "\n",
        "    return sequences, seq_ids, descriptions\n",
        "\n",
        "# Load and inspect sequences\n",
        "if conversion_success:\n",
        "    sequences, seq_ids, descriptions = load_and_inspect_fasta()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "M-Xg1saj6Vg3",
        "outputId": "c4be9781-eed8-446a-e6c3-f0b2954d94a1"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Loading sequences from 18S_sequences.fasta...\n",
            "✓ Loaded 3692 sequences\n",
            "\n",
            "🔍 DETAILED INSPECTION (First 20 sequences):\n",
            "====================================================================================================\n",
            "\n",
            "[1] SEQUENCE ID: NG_065155.1\n",
            "DESCRIPTION: NG_065155.1 Zygosaccharomyces rouxii MUCL 30254 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1801 bp\n",
            "SEQUENCE: TATCTGGTTGATCCTGCCAGTAGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCAATTTATACA...\n",
            "🏷️  EXTRACTED TAXA: Zygosaccharomyces rouxii\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[2] SEQUENCE ID: NG_065153.1\n",
            "DESCRIPTION: NG_065153.1 Stagonospora bicolor ATCC 42652 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1756 bp\n",
            "SEQUENCE: ATATGCTTGTCTCAGATTAAGCCATGCATGTCTAAGTATAAGCAATTATACCGTGAAACTGCGAATGGCTCATTAAATCA...\n",
            "🏷️  EXTRACTED TAXA: Stagonospora bicolor\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[3] SEQUENCE ID: NG_062683.1\n",
            "DESCRIPTION: NG_062683.1 Taphrina populina CBS 337.55 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1744 bp\n",
            "SEQUENCE: CTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCAATTTATACAGTGAAACTGCGAATGGCTCATTAAATCAGT...\n",
            "🏷️  EXTRACTED TAXA: Taphrina populina\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[4] SEQUENCE ID: NG_063110.1\n",
            "DESCRIPTION: NG_063110.1 Epidermophyton floccosum CBS 230.76 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 845 bp\n",
            "SEQUENCE: GCGGTAATTCCAGCTCCAATAGCGTATATTAAAGTTGTTGCAGTTAAAAAGCTCGTAGTTGAACCTTGGGTCTGGCTGGC...\n",
            "🏷️  EXTRACTED TAXA: Epidermophyton floccosum\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[5] SEQUENCE ID: NG_063111.1\n",
            "DESCRIPTION: NG_063111.1 Trichophyton rubrum CBS 392.58 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 844 bp\n",
            "SEQUENCE: GCGGTAATTCCAGCTCCAATAGCGTATATTAAAGTTGTTGCAGTTAAAAAGCTCGTAGTTGAACCTTGGGCCTGGCTGGC...\n",
            "🏷️  EXTRACTED TAXA: Trichophyton rubrum\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[6] SEQUENCE ID: NG_065156.1\n",
            "DESCRIPTION: NG_065156.1 Saccharomycopsis fibuligera MUCL 14481 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1806 bp\n",
            "SEQUENCE: TATCTGGTTGATCCTGCCAGTAGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCAATTATACAG...\n",
            "🏷️  EXTRACTED TAXA: Saccharomycopsis fibuligera\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[7] SEQUENCE ID: NG_061033.1\n",
            "DESCRIPTION: NG_061033.1 Kondoa malvinella IAM 13523 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1757 bp\n",
            "SEQUENCE: AGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCAATCTATACTGTGAAACTGCGAACGGCTCAT...\n",
            "🏷️  EXTRACTED TAXA: Kondoa malvinella\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[8] SEQUENCE ID: NG_061032.1\n",
            "DESCRIPTION: NG_061032.1 Sakaguchia dacryoidea IAM 13522 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 2157 bp\n",
            "SEQUENCE: CTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAACAAATTCATACTGTGAAACTGCGAATGGCTCATTAAATCAG...\n",
            "🏷️  EXTRACTED TAXA: Sakaguchia dacryoidea\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[9] SEQUENCE ID: NG_061105.1\n",
            "DESCRIPTION: NG_061105.1 Hamigera avellanea CBS 295.48 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1770 bp\n",
            "SEQUENCE: TAGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCACTTTATACCGTGAAACTGCGAATGGCTCA...\n",
            "🏷️  EXTRACTED TAXA: Hamigera avellanea\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[10] SEQUENCE ID: NG_062443.1\n",
            "DESCRIPTION: NG_062443.1 Torulaspora microellipsoides CBS 427 18S rRNA gene, partial sequence; from TYPE material >NG_063103.1 Zygotorulaspora mrakii CBS 4218 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1769 bp\n",
            "SEQUENCE: CTGGTTGATCCTGCCAGTAGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCAATTTATACAGTG...\n",
            "🏷️  EXTRACTED TAXA: Torulaspora microellipsoides\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[11] SEQUENCE ID: NG_062140.1\n",
            "DESCRIPTION: NG_062140.1 Kwoniella dendrophila JCM 5357 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1753 bp\n",
            "SEQUENCE: AGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAACGAATTCATACTGTGAAACTGCGAATGGCTCA...\n",
            "🏷️  EXTRACTED TAXA: Kwoniella dendrophila\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[12] SEQUENCE ID: NG_063521.1\n",
            "DESCRIPTION: NG_063521.1 Vishniacozyma globispora JCM 8251 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1753 bp\n",
            "SEQUENCE: AGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAACGAATTCATACTGTGAAACTGCGAATGGCTCA...\n",
            "🏷️  EXTRACTED TAXA: Vishniacozyma globispora\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[13] SEQUENCE ID: NG_063522.1\n",
            "DESCRIPTION: NG_063522.1 Hannaella oryzae JCM 5281 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1755 bp\n",
            "SEQUENCE: AGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAACAAATTCATACTGTGAAACTGCGAATGGCTCA...\n",
            "🏷️  EXTRACTED TAXA: Hannaella oryzae\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[14] SEQUENCE ID: NG_063524.1\n",
            "DESCRIPTION: NG_063524.1 Papiliotrema pseudoalba JCM 5290 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1754 bp\n",
            "SEQUENCE: AGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAACAAATTCATACTGTGAAACTGCGAATGGCTCA...\n",
            "🏷️  EXTRACTED TAXA: Papiliotrema pseudoalba\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[15] SEQUENCE ID: NG_061034.1\n",
            "DESCRIPTION: NG_061034.1 Udeniomyces pyricola JCM 2958 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1759 bp\n",
            "SEQUENCE: AGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAACAAATTTTTACTGTGAAACTGCGAATGGCTCA...\n",
            "🏷️  EXTRACTED TAXA: Udeniomyces pyricola\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[16] SEQUENCE ID: NG_062141.1\n",
            "DESCRIPTION: NG_062141.1 Udeniomyces megalosporus JCM 5269 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1758 bp\n",
            "SEQUENCE: AGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAACAAATTTTTACTGTGAAACTGCGAATGGCTCA...\n",
            "🏷️  EXTRACTED TAXA: Udeniomyces megalosporus\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[17] SEQUENCE ID: NG_063101.1\n",
            "DESCRIPTION: NG_063101.1 Capronia mansonii CBS 101.67 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1731 bp\n",
            "SEQUENCE: GCTTAAGCCATGCATGTCTAAGTATAAGCACTTTATACCGCGAAACTGCGAATGGTTCATTAAATCAGTTATCGTTTATT...\n",
            "🏷️  EXTRACTED TAXA: Capronia mansonii\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[18] SEQUENCE ID: NG_063102.1\n",
            "DESCRIPTION: NG_063102.1 Exophiala exophialae CBS 668.76 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1744 bp\n",
            "SEQUENCE: ATGTCTAAGTATAAGCAATCATACTGTGAAACTGTGAATGGCTCATTAAATCAGTTATCGCTTATTTGATAGTACCTTTA...\n",
            "🏷️  EXTRACTED TAXA: Exophiala exophialae\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[19] SEQUENCE ID: NG_065802.1\n",
            "DESCRIPTION: NG_065802.1 Exophiala phaeomuriformis CBS 131.88 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1743 bp\n",
            "SEQUENCE: CACCAGAAGTTTAAGCCATGCATGTCTAAGTATAAGCACTTTATACCGCGAAACTGCGAATGGCTCATTAAATCAGTTAT...\n",
            "🏷️  EXTRACTED TAXA: Exophiala phaeomuriformis\n",
            "--------------------------------------------------------------------------------\n",
            "\n",
            "[20] SEQUENCE ID: NG_065154.1\n",
            "DESCRIPTION: NG_065154.1 Malbranchea zuffiana UAMH 1875 18S rRNA gene, partial sequence; from TYPE material\n",
            "LENGTH: 1699 bp\n",
            "SEQUENCE: ATGCATGTCTAAGTATAAGCACTATACCGTGAAACTGCGAAGGGCTCATTAAATCAGTTATCGTTTATTTGATAGTACCT...\n",
            "🏷️  EXTRACTED TAXA: Malbranchea zuffiana\n",
            "--------------------------------------------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "headers clearly include binomial names for many entries, so description parsing will yield a sizable labeled subset for supervised PEFT and for validating unsupervised clusters\n",
        "\n"
      ],
      "metadata": {
        "id": "1T7ITuG2_wBC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# CELL 5: Explore Taxonomy SQLite Database\n",
        "\n",
        "def detailed_sqlite_exploration(sqlite_file=\"taxonomy4blast.sqlite3\"):\n",
        "    \"\"\"Detailed exploration of taxonomy SQLite database\"\"\"\n",
        "\n",
        "    if not os.path.exists(sqlite_file):\n",
        "        print(f\"❌ {sqlite_file} not found!\")\n",
        "        return\n",
        "\n",
        "    print(f\"🔍 EXPLORING {sqlite_file}\")\n",
        "    print(\"=\" * 60)\n",
        "\n",
        "    try:\n",
        "        conn = sqlite3.connect(sqlite_file)\n",
        "        cursor = conn.cursor()\n",
        "\n",
        "        # Get database info\n",
        "        cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
        "        tables = [table[0] for table in cursor.fetchall()]\n",
        "        print(f\"📋 TABLES FOUND: {tables}\")\n",
        "\n",
        "        # Explore each table in detail\n",
        "        for table_name in tables:\n",
        "            print(f\"\\n🔍 TABLE: {table_name}\")\n",
        "            print(\"-\" * 40)\n",
        "\n",
        "            # Get column structure\n",
        "            cursor.execute(f\"PRAGMA table_info({table_name})\")\n",
        "            columns = cursor.fetchall()\n",
        "            print(\"COLUMNS:\")\n",
        "            for col in columns:\n",
        "                print(f\"  {col[1]} ({col[2]})\")  # name (type)\n",
        "\n",
        "            # Get row count\n",
        "            cursor.execute(f\"SELECT COUNT(*) FROM {table_name}\")\n",
        "            row_count = cursor.fetchone()[0]\n",
        "            print(f\"ROWS: {row_count}\")\n",
        "\n",
        "            # Get sample data\n",
        "            cursor.execute(f\"SELECT * FROM {table_name} LIMIT 5\")\n",
        "            sample_rows = cursor.fetchall()\n",
        "\n",
        "            if sample_rows:\n",
        "                print(\"SAMPLE DATA:\")\n",
        "                column_names = [col[1] for col in columns]\n",
        "                for i, row in enumerate(sample_rows):\n",
        "                    print(f\"  Row {i+1}:\")\n",
        "                    for col_name, value in zip(column_names, row):\n",
        "                        print(f\"    {col_name}: {value}\")\n",
        "                    print()\n",
        "\n",
        "        conn.close()\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"❌ Error exploring SQLite: {e}\")\n",
        "\n",
        "# Explore taxonomy database\n",
        "detailed_sqlite_exploration()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HtEbd0Lh6bA2",
        "outputId": "b8a6aab8-d88a-4b3e-fc76-6255839070f2"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "🔍 EXPLORING taxonomy4blast.sqlite3\n",
            "============================================================\n",
            "📋 TABLES FOUND: ['TaxidInfo']\n",
            "\n",
            "🔍 TABLE: TaxidInfo\n",
            "----------------------------------------\n",
            "COLUMNS:\n",
            "  taxid (INTEGER)\n",
            "  parent (INTEGER)\n",
            "ROWS: 3037025\n",
            "SAMPLE DATA:\n",
            "  Row 1:\n",
            "    taxid: 1\n",
            "    parent: 1\n",
            "\n",
            "  Row 2:\n",
            "    taxid: 2\n",
            "    parent: 131567\n",
            "\n",
            "  Row 3:\n",
            "    taxid: 6\n",
            "    parent: 335928\n",
            "\n",
            "  Row 4:\n",
            "    taxid: 7\n",
            "    parent: 6\n",
            "\n",
            "  Row 5:\n",
            "    taxid: 9\n",
            "    parent: 32199\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# CELL 6: Extract Labels Using Best Method\n"
      ],
      "metadata": {
        "id": "ma1mCNddAEJd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def extract_comprehensive_labels(seq_ids, descriptions, sqlite_file=\"taxonomy4blast.sqlite3\"):\n",
        "    \"\"\"Extract labels using the most successful method\"\"\"\n",
        "\n",
        "    print(\"🎯 COMPREHENSIVE LABEL EXTRACTION\")\n",
        "    print(\"=\" * 50)\n",
        "\n",
        "    # Method 1: FASTA Description Parsing\n",
        "    print(\"\\n📝 METHOD 1: Parsing FASTA Descriptions\")\n",
        "    desc_labels = {}\n",
        "    genera_found = set()\n",
        "\n",
        "    for seq_id, desc in zip(seq_ids, descriptions):\n",
        "        # Extract scientific names\n",
        "        scientific_match = re.search(r'\\b([A-Z][a-z]+)\\s+([a-z]+)\\b', desc)\n",
        "        if scientific_match:\n",
        "            genus = scientific_match.group(1)\n",
        "            species = scientific_match.group(2)\n",
        "            desc_labels[seq_id] = {\n",
        "                'genus': genus,\n",
        "                'species': species,\n",
        "                'full_name': f\"{genus} {species}\",\n",
        "                'source': 'description'\n",
        "            }\n",
        "            genera_found.add(genus)\n",
        "\n",
        "    print(f\"✓ Extracted from descriptions: {len(desc_labels)} sequences\")\n",
        "    print(f\"✓ Unique genera found: {len(genera_found)}\")\n",
        "\n",
        "    # Method 2: SQLite Database (if available)\n",
        "    sqlite_labels = {}\n",
        "    if os.path.exists(sqlite_file):\n",
        "        print(f\"\\n💾 METHOD 2: SQLite Database Query\")\n",
        "        try:\n",
        "            conn = sqlite3.connect(sqlite_file)\n",
        "            cursor = conn.cursor()\n",
        "\n",
        "            # Try to find the right table and column structure\n",
        "            cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
        "            tables = [t[0] for t in cursor.fetchall()]\n",
        "\n",
        "            for table in tables:\n",
        "                print(f\"  Trying table: {table}\")\n",
        "                try:\n",
        "                    cursor.execute(f\"SELECT * FROM {table} LIMIT 1\")\n",
        "                    sample = cursor.fetchone()\n",
        "                    if sample:\n",
        "                        print(f\"    Sample row: {sample}\")\n",
        "                except:\n",
        "                    continue\n",
        "\n",
        "            conn.close()\n",
        "\n",
        "        except Exception as e:\n",
        "            print(f\"  SQLite query failed: {e}\")\n",
        "\n",
        "    # Combine results\n",
        "    print(f\"\\n📊 FINAL RESULTS:\")\n",
        "    print(f\"Description-based labels: {len(desc_labels)}\")\n",
        "    print(f\"SQLite-based labels: {len(sqlite_labels)}\")\n",
        "\n",
        "    # Use description labels (more reliable for this dataset)\n",
        "    final_labels = desc_labels\n",
        "\n",
        "    # Create summary DataFrame\n",
        "    if final_labels:\n",
        "        label_df = pd.DataFrame([\n",
        "            {\n",
        "                'seq_id': seq_id,\n",
        "                'genus': info['genus'],\n",
        "                'species': info['species'],\n",
        "                'full_name': info['full_name']\n",
        "            }\n",
        "            for seq_id, info in final_labels.items()\n",
        "        ])\n",
        "\n",
        "        print(f\"\\n📋 LABEL SUMMARY:\")\n",
        "        print(f\"Total labeled sequences: {len(label_df)}\")\n",
        "        print(f\"Unique genera: {label_df['genus'].nunique()}\")\n",
        "        print(f\"Most common genera:\")\n",
        "        print(label_df['genus'].value_counts().head(10))\n",
        "\n",
        "        return final_labels, label_df\n",
        "    else:\n",
        "        print(\"❌ No labels extracted successfully\")\n",
        "        return {}, pd.DataFrame()\n",
        "\n",
        "# Extract comprehensive labels\n",
        "if conversion_success and sequences:\n",
        "    final_labels, label_df = extract_comprehensive_labels(seq_ids, descriptions)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Rx9DC4Lg6fxf",
        "outputId": "aeb7980c-90c3-4673-b797-eb5d579696e9"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "🎯 COMPREHENSIVE LABEL EXTRACTION\n",
            "==================================================\n",
            "\n",
            "📝 METHOD 1: Parsing FASTA Descriptions\n",
            "✓ Extracted from descriptions: 3628 sequences\n",
            "✓ Unique genera found: 1433\n",
            "\n",
            "💾 METHOD 2: SQLite Database Query\n",
            "  Trying table: TaxidInfo\n",
            "    Sample row: (1, 1)\n",
            "\n",
            "📊 FINAL RESULTS:\n",
            "Description-based labels: 3628\n",
            "SQLite-based labels: 0\n",
            "\n",
            "📋 LABEL SUMMARY:\n",
            "Total labeled sequences: 3628\n",
            "Unique genera: 1433\n",
            "Most common genera:\n",
            "genus\n",
            "Mucor              49\n",
            "Candida            35\n",
            "Alternaria         31\n",
            "Exophiala          29\n",
            "Ogataea            28\n",
            "Scolecobasidium    27\n",
            "Lophiostoma        24\n",
            "Pichia             22\n",
            "Ophiocordyceps     22\n",
            "Aspergillus        22\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Strong result—3,628 of 3,692 sequences now have parsed genus/species labels, giving a large supervised set plus a small unlabeled remainder for discovery and semi‑supervised loops."
      ],
      "metadata": {
        "id": "4dsLYqRFAXVf"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Build supervised training targets at multiple ranks by collapsing the parsed binomials to genus, family, order, etc., after mapping names to taxids and ranks; this enables macro‑F1 at genus/species and robustness at higher ranks"
      ],
      "metadata": {
        "id": "FNyy1W66AfDg"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Retain 64 unlabeled sequences for the unsupervised branch; they can seed novel‑taxon clusters with UMAP→HDBSCAN and later be pseudo‑labeled via cluster consensus BLAST or majority label propagation"
      ],
      "metadata": {
        "id": "yNyHCdxMAiZA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print(\"\\n\" + \"=\"*60)\n",
        "print(\"🎉 BLAST CONVERSION & LABEL EXTRACTION COMPLETE!\")\n",
        "print(\"=\"*60)\n",
        "\n",
        "if conversion_success:\n",
        "    print(f\"✅ FASTA Conversion: SUCCESS\")\n",
        "    print(f\"   📁 Output file: 18S_sequences.fasta\")\n",
        "    print(f\"   📊 Total sequences: {len(sequences)}\")\n",
        "    print(f\"   📏 Average sequence length: {sum(len(s) for s in sequences)/len(sequences):.1f} bp\")\n",
        "\n",
        "    if final_labels:\n",
        "        print(f\"✅ Label Extraction: SUCCESS\")\n",
        "        print(f\"   🏷️  Labeled sequences: {len(final_labels)}\")\n",
        "        print(f\"   🧬 Unique genera: {len(set(info['genus'] for info in final_labels.values()))}\")\n",
        "        print(f\"   📈 Label coverage: {len(final_labels)/len(sequences)*100:.1f}%\")\n",
        "\n",
        "        print(f\"\\n🔬 EXAMPLE EXTRACTED DATA:\")\n",
        "        for i, (seq_id, seq) in enumerate(zip(seq_ids[:3], sequences[:3])):\n",
        "            print(f\"\\nSequence {i+1}:\")\n",
        "            print(f\"  ID: {seq_id}\")\n",
        "            print(f\"  Length: {len(seq)} bp\")\n",
        "            if seq_id in final_labels:\n",
        "                print(f\"  Genus: {final_labels[seq_id]['genus']}\")\n",
        "                print(f\"  Species: {final_labels[seq_id]['species']}\")\n",
        "            else:\n",
        "                print(f\"  Label: Not found\")\n",
        "            print(f\"  Sequence: {seq[:60]}...\")\n",
        "    else:\n",
        "        print(f\"❌ Label Extraction: FAILED\")\n",
        "        print(\"   Try manual inspection of file descriptions\")\n",
        "else:\n",
        "    print(f\"❌ FASTA Conversion: FAILED\")\n",
        "    print(\"   Check if all required .ndb, .nhr, .nin, .nsq files are uploaded\")\n",
        "\n",
        "print(f\"\\n🚀 READY FOR NEXT STEPS:\")\n",
        "if conversion_success and final_labels:\n",
        "    print(\"   ✅ You can now proceed with UMAP/HDBSCAN clustering\")\n",
        "    print(\"   ✅ You can proceed with LoRA fine-tuning\")\n",
        "    print(f\"   📊 You have {len(final_labels)} labeled sequences for supervised learning\")\n",
        "    print(f\"   🔍 You have {len(sequences) - len(final_labels)} unlabeled sequences for novel discovery\")\n",
        "elif conversion_success:\n",
        "    print(\"   ✅ You can proceed with unsupervised UMAP/HDBSCAN only\")\n",
        "    print(\"   ❌ No supervised learning possible without labels\")\n",
        "else:\n",
        "    print(\"   ❌ Fix file upload and conversion issues first\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "giw_2SAz6oCK",
        "outputId": "d9fbc640-8dc1-4cdd-86c4-4f818b073dd9"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "============================================================\n",
            "🎉 BLAST CONVERSION & LABEL EXTRACTION COMPLETE!\n",
            "============================================================\n",
            "✅ FASTA Conversion: SUCCESS\n",
            "   📁 Output file: 18S_sequences.fasta\n",
            "   📊 Total sequences: 3692\n",
            "   📏 Average sequence length: 1333.0 bp\n",
            "✅ Label Extraction: SUCCESS\n",
            "   🏷️  Labeled sequences: 3628\n",
            "   🧬 Unique genera: 1433\n",
            "   📈 Label coverage: 98.3%\n",
            "\n",
            "🔬 EXAMPLE EXTRACTED DATA:\n",
            "\n",
            "Sequence 1:\n",
            "  ID: NG_065155.1\n",
            "  Length: 1801 bp\n",
            "  Genus: Zygosaccharomyces\n",
            "  Species: rouxii\n",
            "  Sequence: TATCTGGTTGATCCTGCCAGTAGTCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCT...\n",
            "\n",
            "Sequence 2:\n",
            "  ID: NG_065153.1\n",
            "  Length: 1756 bp\n",
            "  Genus: Stagonospora\n",
            "  Species: bicolor\n",
            "  Sequence: ATATGCTTGTCTCAGATTAAGCCATGCATGTCTAAGTATAAGCAATTATACCGTGAAACT...\n",
            "\n",
            "Sequence 3:\n",
            "  ID: NG_062683.1\n",
            "  Length: 1744 bp\n",
            "  Genus: Taphrina\n",
            "  Species: populina\n",
            "  Sequence: CTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCAATTTATACAGTGAAACTGC...\n",
            "\n",
            "🚀 READY FOR NEXT STEPS:\n",
            "   ✅ You can now proceed with UMAP/HDBSCAN clustering\n",
            "   ✅ You can proceed with LoRA fine-tuning\n",
            "   📊 You have 3628 labeled sequences for supervised learning\n",
            "   🔍 You have 64 unlabeled sequences for novel discovery\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "this dataset is ready for both supervised PEFT fine-tuning and unsupervised discovery with UMAP→HDBSCAN."
      ],
      "metadata": {
        "id": "rcgVVZXDA1ZM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Save processed data\n",
        "if conversion_success:\n",
        "    # Save sequences as CSV for easy loading\n",
        "    df = pd.DataFrame({\n",
        "        'seq_id': seq_ids,\n",
        "        'sequence': sequences,\n",
        "        'description': descriptions,\n",
        "        'sequence_length': [len(s) for s in sequences]\n",
        "    })\n",
        "\n",
        "    # Add labels if available\n",
        "    if final_labels:\n",
        "        df['genus'] = df['seq_id'].map(lambda x: final_labels.get(x, {}).get('genus', 'Unknown'))\n",
        "        df['species'] = df['seq_id'].map(lambda x: final_labels.get(x, {}).get('species', 'Unknown'))\n",
        "        df['has_label'] = df['seq_id'].map(lambda x: x in final_labels)\n",
        "\n",
        "    # Save to CSV\n",
        "    df.to_csv('edna_sequences_processed.csv', index=False)\n",
        "    print(f\"✅ Saved processed data to 'edna_sequences_processed.csv'\")\n",
        "    print(f\"📊 Dataset shape: {df.shape}\")\n",
        "\n",
        "    # Show data preview\n",
        "    print(f\"\\n📋 DATA PREVIEW:\")\n",
        "    print(df.head())\n",
        "\n",
        "    if final_labels:\n",
        "        print(f\"\\n🏷️  LABEL DISTRIBUTION:\")\n",
        "        label_counts = df[df['has_label']]['genus'].value_counts().head(10)\n",
        "        print(label_counts)\n",
        "\n",
        "print(f\"\\n🎯 NEXT STEPS:\")\n",
        "print(\"1. Run UMAP + HDBSCAN on the sequences for novel taxa discovery\")\n",
        "print(\"2. If you have labels, set up LoRA fine-tuning\")\n",
        "print(\"3. Combine results for comprehensive biodiversity assessment\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4PYD0LUu6tfh",
        "outputId": "f5971993-a578-45eb-cd49-c9d280d1c087"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "✅ Saved processed data to 'edna_sequences_processed.csv'\n",
            "📊 Dataset shape: (3692, 7)\n",
            "\n",
            "📋 DATA PREVIEW:\n",
            "        seq_id                                           sequence  \\\n",
            "0  NG_065155.1  TATCTGGTTGATCCTGCCAGTAGTCATATGCTTGTCTCAAAGATTA...   \n",
            "1  NG_065153.1  ATATGCTTGTCTCAGATTAAGCCATGCATGTCTAAGTATAAGCAAT...   \n",
            "2  NG_062683.1  CTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTATAAGCAATTTA...   \n",
            "3  NG_063110.1  GCGGTAATTCCAGCTCCAATAGCGTATATTAAAGTTGTTGCAGTTA...   \n",
            "4  NG_063111.1  GCGGTAATTCCAGCTCCAATAGCGTATATTAAAGTTGTTGCAGTTA...   \n",
            "\n",
            "                                         description  sequence_length  \\\n",
            "0  NG_065155.1 Zygosaccharomyces rouxii MUCL 3025...             1801   \n",
            "1  NG_065153.1 Stagonospora bicolor ATCC 42652 18...             1756   \n",
            "2  NG_062683.1 Taphrina populina CBS 337.55 18S r...             1744   \n",
            "3  NG_063110.1 Epidermophyton floccosum CBS 230.7...              845   \n",
            "4  NG_063111.1 Trichophyton rubrum CBS 392.58 18S...              844   \n",
            "\n",
            "               genus    species  has_label  \n",
            "0  Zygosaccharomyces     rouxii       True  \n",
            "1       Stagonospora    bicolor       True  \n",
            "2           Taphrina   populina       True  \n",
            "3     Epidermophyton  floccosum       True  \n",
            "4       Trichophyton     rubrum       True  \n",
            "\n",
            "🏷️  LABEL DISTRIBUTION:\n",
            "genus\n",
            "Mucor              49\n",
            "Candida            35\n",
            "Alternaria         31\n",
            "Exophiala          29\n",
            "Ogataea            28\n",
            "Scolecobasidium    27\n",
            "Lophiostoma        24\n",
            "Pichia             22\n",
            "Ophiocordyceps     22\n",
            "Aspergillus        22\n",
            "Name: count, dtype: int64\n",
            "\n",
            "🎯 NEXT STEPS:\n",
            "1. Run UMAP + HDBSCAN on the sequences for novel taxa discovery\n",
            "2. If you have labels, set up LoRA fine-tuning\n",
            "3. Combine results for comprehensive biodiversity assessment\n"
          ]
        }
      ]
    }
  ]
}